packages/flutter/lib/src/widgets/app.dart

```diff
@@ -439,29 +454,42 @@ class WidgetsApp extends StatefulWidget {
   /// {@endtemplate}
   final Iterable<LocalizationsDelegate<dynamic>> localizationsDelegates;
 
-  /// {@template flutter.widgets.widgetsApp.localeResolutionCallback}
+  /// {@template flutter.widgets.widgetsApp.localeListResolutionCallback}
   /// This callback is responsible for choosing the app's locale
   /// when the app is started, and when the user changes the
   /// device's locale.
   ///
-  /// The returned value becomes the locale of this app's [Localizations]
-  /// widget. The callback's `locale` parameter is the device's locale when
-  /// the app started, or the device locale the user selected after the app was
-  /// started. The callback's `supportedLocales` parameter is just the value
-  /// [supportedLocales].
+  /// When a [localeListResolutionCallback] is provided, Flutter will first attempt to
+  /// resolve the locale with the provided [localeListResolutionCallback]. If the
+  /// callback or result is null, will fallback to trying the [localeResolutionCallback].
+  /// If both [localeResolutionCallback] and [localeListResolutionCallback] are left null
+  /// or fail to resolve (return null), the [WidgetsApp.fallbackLocaleResolution]
+  /// fallback algorithm will be used.
   ///
-  /// If the callback is null or if it returns null then the resolved locale is:
+  /// The priority of each available fallback is:
   ///
-  /// - The callback's `locale` parameter if it's equal to a supported locale.
-  /// - The first supported locale with the same [Locale.languageCode] as the
-  ///   callback's `locale` parameter.
-  /// - The first locale in [supportedLocales].
+  ///  1. [localeListResolutionCallback] is attempted first.
+  ///  2. [localeResolutionCallback] is attempted second.
+  ///  3. Flutter's [WidgetsApp.fallbackLocaleResolution] algorithm is attempted last.
   /// {@endtemplate}
   ///
+  /// This callback considers the entire list of preferred locales.
+  ///
   /// See also:
   ///
   ///  * [MaterialApp.localeResolutionCallback], which sets the callback of the
   ///    [WidgetsApp] it creates.
+  final LocaleListResolutionCallback localeListResolutionCallback;
+
+  /// {@macro flutter.widgets.widgetsApp.localeListResolutionCallback}
+  ///
+  /// This callback considers only the default locale, which is the first locale
+  /// in the preferred locales list.
+  ///
+  /// See also:
+  ///
+  ///  * [MaterialApp.localeListResolutionCallback], which sets the callback of the
+  ///    [WidgetsApp] it creates.
```

Comment:
How do I decide if I should provide a localeListResolutionCallback or localeResolutionCallback or both? We should document that in an obvious place.

Reply:
This is here primarily for backwards compatibility. If an existing algorithm matches on just the top locale, use the legacy `localeResolutionCallback`. Otherwise, use the list callback that supports the full preferred locale list. I can document that the localeListResolutionCallback is preferred for new implementations. `localeListResolutionCallback` can do everything `localeResolutionCallback` can do.